pkg/vpcmodel/connectivityAbstraction.go

```diff
@@ -0,0 +1,159 @@
+//      nodeSetToNodeSet: connections of   <node     in the nodeSet>  ->  <node     in the nodeSet>
+//      otherFromNodeSet: connections of   <node     in the nodeSet>  ->  <node not in the nodeSet>
+//      otherToNodeSet:   connections of   <node not in the nodeSet>  ->  <node     in the nodeSet>
+// 2. for the last three groups, we check the abstraction assumption holds
+// 3. we do the abstraction ( for now, even if the abstraction assumption does not hold):
+// the connectivity of N->NS is union of all of N->N1, N->N2, N->N3 ...
+// todo: what to do if the abstraction assumption does not hold?
+
+func nodeSetConnectivityAbstraction(nodesConn GeneralConnectivityMap, nodeSet NodeSet) GeneralConnectivityMap {
+	otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet := splitConnectivityByNodeSet(nodesConn, nodeSet)
+	checkConnectivityAbstractionValidity(otherFromNodeSet, nodeSet, false)
+	checkConnectivityAbstractionValidity(otherToNodeSet, nodeSet, true)
+	checkConnectivityAbstractionValidity(nodeSetToNodeSet, nodeSet, true)
+	return mergeConnectivityWithNodeSetAbstraction(otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet, nodeSet)
+}
+
+// splitConnectivityByNodeSet() split the connectivity to the four groups
+// each group is kept as GeneralConnectivityMap.
+// usually, GeneralConnectivityMap is a map form src to dst.
+// however, the third group is hold as a map from dst to src (and therefore called otherFromNodeSet and not nodeSetToOther)
+// see the reason on mergeConnectivityWithNodeSetAbstraction()
+
+func splitConnectivityByNodeSet(nodesConn GeneralConnectivityMap, nodeSet NodeSet) (
+	otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet GeneralConnectivityMap) {
+	otherToOther = GeneralConnectivityMap{}
+	nodeSetToNodeSet = GeneralConnectivityMap{}
+	otherFromNodeSet = GeneralConnectivityMap{}
+	otherToNodeSet = GeneralConnectivityMap{}
+	for src, nodeConns := range nodesConn {
+		for dst, conns := range nodeConns {
+			srcNode, srcIsNode := src.(Node)
+			dstNode, dstIsNode := dst.(Node)
+			srcInSet := srcIsNode && slices.Contains(nodeSet.Nodes(), srcNode)
+			dstInSet := dstIsNode && slices.Contains(nodeSet.Nodes(), dstNode)
+			switch {
+			case (!srcInSet && !dstInSet) || conns.IsEmpty():
+				otherToOther.updateAllowedConnsMap(src, dst, conns)
+			case srcInSet && dstInSet:
+				nodeSetToNodeSet.updateAllowedConnsMap(src, dst, conns)
+			case srcInSet && !dstInSet:
+				otherFromNodeSet.updateAllowedConnsMap(dst, src, conns)
+			case !srcInSet && dstInSet:
+				otherToNodeSet.updateAllowedConnsMap(src, dst, conns.Copy())
+			}
+		}
+	}
+	return otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet
+}
+
+// checkConnectivityAbstractionValidity() checks if the abstraction assumption holds
+// it does it on each group separately.
+// for now it creates a string
+// todo: - how to report this string? what format?
+// for now, No need to CR it, we do not do anything with the result
```

Comment:
Was not reviewed in this stage
Perhaps better not merge the code in this stage - maybe comment it, so that once it is ready for review the review is easier - so that not only the deltas are presented

Reply:
commented out